File: software/examples/delphi/ExamplePixelMatrix.pas
program ExamplePixelMatrix;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, Math, IPConnection, BrickletOLED64x48;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'ABC2'; { Change to your UID }
  SCREEN_WIDTH = 64;
  SCREEN_HEIGHT = 48;

type
  TPixelMatrix = array[0..(SCREEN_HEIGHT - 1), 0..(SCREEN_WIDTH - 1)] of boolean;
  TExample = class
  private
    ipcon: TIPConnection;
    oled: TBrickletOLED64x48;
    procedure DrawMatrix(pixels: TPixelMatrix);
  public
    procedure Execute;
end;

var
  e: TExample;

procedure TExample.DrawMatrix(pixels: TPixelMatrix);
var
  i, j, k: integer;
  page: byte;
  column: array[0..5, 0..(SCREEN_WIDTH - 1)] of byte;
begin
  for i := 0 to 5 do
  begin
    for j := 0 to SCREEN_WIDTH - 1 do
    begin
      page := 0;
      for k := 0 to 7 do begin
        if (pixels[(i*8) + k, j]) then begin
          page := page or (1 << k);
        end;
      end;

      column[i][j] := page;
    end;
  end;
  oled.NewWindow(0, SCREEN_WIDTH - 1, 0, 5);
  for i := 0 to 5 do begin
    oled.write(column[i]);
  end;
end;

procedure TExample.Execute;
var
  pixel_matrix: TPixelMatrix;
  w, h, p: integer;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  oled := TBrickletOLED64x48.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Clear display }
  oled.ClearDisplay;

  { Pixel matrix with all pixels turned off }
  for h := 0 to SCREEN_HEIGHT - 1 do begin
    for w := 0 to SCREEN_WIDTH - 1 do begin
      pixel_matrix[h, w] := false;
    end;
  end;

  { Draw check pattern }
  for w := 0 to SCREEN_WIDTH - 1 do begin
    for h := 0 to SCREEN_HEIGHT - 1 do begin
      if (Floor(w/5) mod 2 = 0) then begin
        pixel_matrix[h, w] := true;
      end;
      if (Floor(h/5) mod 2 = 0) then begin
        pixel_matrix[h, w] := true;
      end;
    end;
  end;

  e.DrawMatrix(pixel_matrix);

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/vbnet/ExamplePixelMatrix.vb
Imports System
Imports System.Math
Imports Tinkerforge

Module ExamplePixelMatrix
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change XYZ to the UID of your OLED 64x48 Bricklet
    Const WIDTH As Integer = 64
    Const HEIGHT As Integer = 48

    Sub DrawMatrix(ByRef oled As BrickletOLED64x48, ByVal pixels()() As Boolean)
        Dim pages()() As Byte = New Byte(HEIGHT \ 8)() {}

        For row As Integer = 0 To HEIGHT \ 8 - 1
            pages(row) = New Byte(WIDTH) {}

            For column As Integer = 0 To WIDTH - 1
                pages(row)(column) = 0

                For bit As Integer = 0 To 7
                    If pixels((row * 8) + bit)(column) Then
                        pages(row)(column) = pages(row)(column) Or Convert.ToByte(1 << bit)
                    End If
                Next bit
            Next column
        Next row

        oled.NewWindow(0, WIDTH - 1, 0, HEIGHT \ 8 - 1)

        For row As Integer = 0 To HEIGHT \ 8 - 1
            oled.Write(pages(row))
        Next
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim oled As New BrickletOLED64x48(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Clear display
        oled.ClearDisplay()

        ' Draw checkerboard pattern
        Dim pixels()() As Boolean = New Boolean(HEIGHT)() {}

        For row As Integer = 0 To HEIGHT - 1
            pixels(row) = New Boolean(WIDTH) {}

            For column As Integer = 0 To WIDTH - 1
                pixels(row)(column) = (row \ 8) Mod 2 = (column \ 8) Mod 2
            Next row
        Next column

        DrawMatrix(oled, pixels)

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module
